server/routerlicious/packages/services-client/src/error.ts

```diff
@@ -3,15 +3,186 @@
+ * over the network. Network communication is subject to a diverse range of errors. {@link NetworkError} helps
+ * convey more information than a simple HTTP status code, allowing services to be aware of the context of a
+ * network error and making those services more prepared to react to such kinds of errors.
+ */
 export class NetworkError extends Error {
     constructor(
         /**
          * HTTP status code that describes the error.
+         * @public
          */
         public readonly code: number,
+        /**
+         * The message associated with the error.
+         * @public
+         */
         message: string,
+        /**
+         * Optional boolean indicating whether this is an error that can be retried.
+         * Only relevant when {@link NetworkError#isFatal} is false.
+         * @public
+         */
+        public readonly canRetry?: boolean,
+        /**
+         * Optional boolean indicating whether this error is fatal. This generally indicates that the error causes
+         * negative, non-recoverable impact to the component/caller and cannot be ignored.
+         * @public
+         */
+        public readonly isFatal?: boolean,
+        /**
+         * Optional value representing the time in milliseconds that should be waited before retrying.
+         * @public
+         */
+        public readonly retryAfterMs?: number,
     ) {
         super(message);
         this.name = "NetworkError";
     }
+
+    /**
+     * Gets the details associated with this {@link NetworkError}.
+     * @returns A simple string conveying the message if no other details are included in this {@link NetworkError},
+     * or an {@link INetworkErrorDetails} object otherwise.
+     * @public
+     */
+    public get details(): INetworkErrorDetails | string {
+        if (this.canRetry === undefined && this.isFatal === undefined && this.retryAfterMs === undefined) {
+            return this.message;
+        }
+
+        return {
+            message: this.message,
+            canRetry: this.canRetry,
+            isFatal: this.isFatal,
+            retryAfter: this.retryAfterMs,
+        };
+    }
+}
+
+/**
+ * Convenience function for generating a {@link NetworkError}.
+ * @remarks Generates a {@link NetworkError} instance appropriately configured given the status code and error data
+ * provided. This function is intended to be used in situations where a {@link NetworkError} is dynamically created
+ * based variable parameters. That is, when it is not known whether the status code can be 404 or 500.
+ * @param statusCode - HTTP status code that describes the error.
+ * @param errorData - Optional additional data associated with the error. Can either be a simple string representing
+ *  the message, or an {@link INetworkErrorDetails} object.
+ * @returns A simple string conveying the message if no other details are included in this {@link NetworkError},
```

Comment:
nit: it looks like it always throws a NetworkError, never a string

Reply:
Nice catch! It was a copy/paste failure from my side :) let me fix that.